image_view/src/nodes/video_recorder.cpp

```diff
@@ -34,31 +36,73 @@ ros::Time g_last_wrote_time = ros::Time(0);
 std::string encoding;
 std::string codec;
 int fps;
-std::string filename;
+std::string base_filename, filename;
 double min_depth_range;
 double max_depth_range;
 bool use_dynamic_range;
 int colormap;
-
+bool stamped_filename;
+bool rolling_buffer;
+ros::Duration rb_video_lenght;
+int rb_n_videos;
+ros::Time video_creation_stamp;
+std::vector<std::string> filenames;
```

Comment:
I'm a super huge not-fan of this. This is a global variable of variable size. The other variables while changing probably cannot cause additional memory allocation, but this will constantly. This shouldn't be global or should be set at runtime to a constant unchanging size.

Reply:
Sorry, but my c++ skills are pretty basic and I'm not sure of how to exactly do what you suggest. I would need some help on this.

I guess that it would it be enough to just do `filenames.reserve(rb_n_videos + 1)` once we read the `rb_n_videos` parameter for the `std:vector`. However, there is no way (that I know of) of reserving also the size of the stored `std::string`, right?

Second option that comes to my mind is to use something like a non global `std::array<char[filename.size()], rb_n_videos>  filenames` and use it as input/output parameter to `generateStampedFilename` function. But I really need to check if this can be done (and how).

Anyway (and probably I'll be beaten up for this comment), in the worst case you would have `rb_n_videos + 1` memory allocations... Is it really so much of a big deal? :sweat_smile: I though this was exactly the purpose of the `vector` type.

Reply:
@IvanVN - Just a couple of suggestions/question here:

1. How are you actually using this filenames list? It looks like you use a function called `remove()` on the first filename in the vector but I don't exactly know what this function is doing or where it is included.
2. This should probably be a `queue` instead of a `vector`. A vector is not designed for efficiently removing items from the front of the list. A `queue` is exactly for this - it's an implementation of a FIFO (first-in-first-out) buffer.
3. If you reserve the amount of memory ahead of time equal to `rb_n_videos`, then that's the majority of the allocation done up-front. No, it won't reserve the size of each `std::string` but a vector is essentially a list of pointers to the actual storage locations for the elements anyway so the `reserve(N)` function will reserve the memory for one pointer for each element N. The pointed-to elements may change in size or location but the allocated memory for each pointer remains the same unless you add N+1 elements.
4. Rather than allocating `rb_n_videos` + 1, you could check the size of the container before adding the new element.

Reply:
All above work. I know this pipeline may be used on some of the delivery robots and those guys are more concerned about that type of thing than your typical indoor robot (but far less than the autonomous car people). Just making sure we're not introducing a changing memory allocation where there wasn't one before would be good. The other global variables are statics so on runtime they allocate once and just hang out, this buffer changes size which is what I'm wanting to resolve.